### PR TITLE
[xDS Proto] Modernize buildgen scripts

### PR DESCRIPTION
**Pull request by lidizheng**
Related #25272, based on #25510.

Adding type annotations, modernize Python API usages, and improve debuggability.

### CODE REVIEW

File: tools/buildgen/generate_build_additions.sh
Comment (jtattermusch) [inline tools/buildgen/generate_build_additions.sh:34]: nit: why this change?

File: tools/run_tests/python_utils/jobset.py
Comment (jtattermusch) [inline tools/run_tests/python_utils/jobset.py:280]: did we see this failure before?  how does it exhibit itself? is there a link with more info about this issue?

File: tools/buildgen/plugins/expand_filegroups.py
Comment (jtattermusch) [inline tools/buildgen/plugins/expand_filegroups.py:50]: btw, I think since the migration to extract_metadata_from_bazel.py, we might no longer have any filegroups - it might be worth considering getting rid of this "plugin".

**Review by jtattermusch — COMMENTED** [review]
This is looking good (see my comments), but I'd like @gnossen  to review the python code more in detail.

File: tools/buildgen/generate_build_additions.sh
Comment (lidizheng) [reply tools/buildgen/generate_build_additions.sh:34]: We have 12 gen_build_yaml script in our repo, this will help us identify which intermediate file is generated by which script. E.g., `src/re2/gen_build_yaml.py` generates `gen_src_re2_gen_build_yaml_py.yaml.12345` instead of `gen12345`.

File: tools/run_tests/python_utils/jobset.py
Comment (lidizheng) [reply tools/run_tests/python_utils/jobset.py:280]: It occurs to me on my MacBook. I tried to twitch file handle limit, or remove temporary files. But it's unclear why the `os.unlink` failed with a system error.

File: tools/buildgen/plugins/expand_filegroups.py
Comment (lidizheng) [reply tools/buildgen/plugins/expand_filegroups.py:50]: After removing the `tools/buildgen/plugins/expand_filegroups.py` script, plugin `tools/buildgen/plugins/make_fuzzer_tests.py` failed.

Comment (jtattermusch) [reply tools/buildgen/plugins/expand_filegroups.py:50]: sg.

File: tools/run_tests/python_utils/jobset.py
Comment (jtattermusch) [reply tools/run_tests/python_utils/jobset.py:280]: Right, I'm getting that too when I run `tools/buildgen/generate_projects.sh`
```
  File "tools/buildgen/generate_projects.py", line 98, in <module>
    err_cnt, _ = jobset.run(jobs, maxjobs=args.jobs)
  File "tools/buildgen/../run_tests/python_utils/jobset.py", line 589, in run
    if not js.start(cmdline):
  File "tools/buildgen/../run_tests/python_utils/jobset.py", line 474, in start
    self._quiet_success)
  File "tools/buildgen/../run_tests/python_utils/jobset.py", line 266, in __init__
    self.start()
  File "tools/buildgen/../run_tests/python_utils/jobset.py", line 280, in start
    self._logfile = tempfile.TemporaryFile()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/tempfile.py", line 624, in TemporaryFile
    _os.unlink(name)
```

But when I tried your fix, I'm now getting a warning:
```
Exception ignored in: <bound method _TemporaryFileCloser.__del__ of <tempfile._TemporaryFileCloser object at 0x7fc497733e10>>
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/tempfile.py", line 452, in __del__
    self.close()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/tempfile.py", line 448, in close
    unlink(self.name)
InterruptedError: [Errno 4] Interrupted system call: '/var/folders/7w/1rl183bs1cz4ghlh8d6mnfhm006nph/T/tmp6_4e_41_'
```
The buildgen then passes, so there is an improvement, but it's not a 100% fix.

**Review by jtattermusch — APPROVED** [review]
LGTM once @gnossen  approves the changes to python code

**Comment by lidizheng** [comment]
@gnossen PTAL at the changes to the Python scripts

File: tools/buildgen/_utils.py
Comment (gnossen) [inline tools/buildgen/_utils.py:37]: Nit: `Mapping` would be more general.

File: tools/buildgen/generate_projects.py
Comment (gnossen) [inline tools/buildgen/generate_projects.py:81]: Debug print?

File: tools/buildgen/generate_projects.py
Comment (gnossen) [inline tools/buildgen/generate_projects.py:93]: Nit: `sorted` has a `reverse` kwarg.

File: tools/buildgen/generate_projects.py
Comment (gnossen) [inline tools/buildgen/generate_projects.py:97]: `os.path.join`?

**Review by gnossen — APPROVED** [review]
Nice improvement!

File: tools/buildgen/_utils.py
Comment (lidizheng) [reply tools/buildgen/_utils.py]: Replaced with `Mapping`.

File: tools/buildgen/generate_projects.py
Comment (lidizheng) [reply tools/buildgen/generate_projects.py]: Good catch!

File: tools/buildgen/generate_projects.py
Comment (lidizheng) [reply tools/buildgen/generate_projects.py]: Updated. TIL.

File: tools/buildgen/generate_projects.py
Comment (lidizheng) [reply tools/buildgen/generate_projects.py]: Good catch!

**Pull request merged**
